### PR TITLE
introduce patch(.diff) to slstatus and dmenu

### PR DESCRIPTION
**Pull request by ahdyt**
###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [x] macOS
   - [x] other Linux distributions
- [x] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [x] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by fnlaai** [comment]
@oxzi Alv, I'm introduce patches, I already tested with sudo nixos-rebuild switch -I nixpkgs=/cloned/github/nixpkgs, and it works.

File: pkgs/top-level/all-packages.nix
Comment (NobbZ) [inline pkgs/top-level/all-packages.nix:20432]: The `dmenu` package usually has a default of `[ ./xim.patch ]` in it, you are basically removing it here.

Is this intentional?

File: pkgs/top-level/all-packages.nix
Comment (NobbZ) [inline pkgs/top-level/all-packages.nix:20435]: The function returned from that file doesn't even have `patches` in it's argset, are you sure this works?

I haven't tested it though, as I'm doing the comments from a mobile.

**Review by NobbZ — COMMENTED** [review]
On top of the 2 comments I have left inline, I do not like that the config from the module system are used like this, I think instead the module system should use `override`/`overrideAttrs` to apply patches and changes.

Though in the chat I was told that this PR is based on prior art with other "suckless" software packages.

As this review is solely based on personal opinion and reading the diff/referred files and not actually running any builds that would verify buildability or functionality of these changes, this is posted as "comment"

File: pkgs/top-level/all-packages.nix
Comment (NobbZ) [reply pkgs/top-level/all-packages.nix:20435]: Just see it in the error of the GitHub action...

It does indeed not evaluate because of this:

https://gist.github.com/GrahamcOfBorg/0f56c0ebd42091febfe5410583975bf6#file-output-path-comparison-L9

**Comment by fnlaai** [comment]
I'm done here, although I can build derivation on my own, so get lost those nix newcomer who will follow pattern nixpkgs.config.<suckless's-packages>.patches ! make your own overlays/derivation~

**Comment by NobbZ** [comment]
Sorry if I offended you, that was not my intend.

I was just asking questions, about how you do things, and if you realized how your changes might affect others, as you overwrite current default behaviour with the patch for `dmenu` beeing basically "deleted".

Also I pointed out that another of your changes does not build at all.

Last but not least, I realise that your work is based on prior art and because of that has to be considered "okay", even though I do not agree with the prior art.

If you want to, you can totally ignore my comment about not liking the way how the changes use the modules, but still, the breaking build of `dmenu-wayland` remains and my question about your intend of dropping the `./xim.patch` for `dmenu`.

I'm totally fine with the intend of making those software easier to configure via modules, though not at the cost of failing builds or unintentionally dropped patches.

So please fix the build for `dmenu-wayland`, or the PR can't be merged, as its not even passing simple evaluation checks. This is the bare minimum to get a proper and meaningful review of someone more experienced with the suckless ecosystem and their integration into nixos/nixpkgs.

**Comment by fnlaai** [comment]
No, it's my fault, I have no idea what xim patch do, and the wayland one, thought it was take patches args, ah.

**Comment by NobbZ** [comment]
I think we should ping for a review:

* @oxzi as the maintainer of `slstatus`
* @pSub as the maintainer of `dmenu`

File: pkgs/applications/misc/dmenu/default.nix
Comment (oxzi) [inline pkgs/applications/misc/dmenu/default.nix:1]: This became obsolete after #98267 landed. Thus, the PR should be rebased.

File: pkgs/top-level/all-packages.nix
Comment (oxzi) [inline pkgs/top-level/all-packages.nix:7168]: Looks good to me.

File: pkgs/top-level/all-packages.nix
Comment (oxzi) [reply pkgs/top-level/all-packages.nix:20432]: This old default patch was removed by bumping to 5.0 in #98267. Now the same `patches` code can be used as for `slstatus` above.

**Review by oxzi — CHANGES_REQUESTED** [review]
Please excuse my late response, @h1dyt. Everything looks good to me, except the legacy `dmenu` changes which became obsolete after #98267. Could you please rebase your PR? Thanks!

**Comment by h1dyt** [comment]
I've updated my repo but it's kinda adding new "commits" ? is that how to rebase? I guess you've should to squash it.

File: pkgs/top-level/all-packages.nix
Comment (oxzi) [inline pkgs/top-level/all-packages.nix:20656]: The `xim` patch became obsolete. Thus, you should change this to an empty list.

**Review by oxzi — CHANGES_REQUESTED** [review]
Sorry, I missed the faulty default in the `all-packages.nix` before. Could you please change this one again? Thanks a lot!

**Review by ahdyt — COMMENTED** [review]
> Sorry, I missed the faulty default in the `all-packages.nix` before. Could you please change this one again? Thanks a lot!

yes, ofborg reminds me.

**Review by oxzi — APPROVED** [review]
Looks good to me (:

**Pull request closed without merging**
